File: proto/test/Test/Canonical.hs
-- | The canonical JSON printer: exact bytes on fixed cases, and a round trip
-- on generated values (encode, parse with aeson, re-encode: same bytes; and
-- the parse is the original value).
module Test.Canonical (tests) where

import Data.Aeson (Value (..), decodeStrict)
import qualified Data.Aeson.Key as Key
import qualified Data.Aeson.KeyMap as KeyMap
import qualified Data.ByteString.Char8 as BC
import qualified Data.Text as T
import qualified Data.Vector as V
import Rue.Proto.Json.Canonical (encode)
import Test.Tasty (TestTree, testGroup)
import Test.Tasty.HUnit (testCase, (@?=))
import Test.Tasty.QuickCheck

tests :: TestTree
tests =
  testGroup
    "canonical-json"
    [ testCase "empty object" $ encode (Object mempty) @?= Right (BC.pack "{}\n")
    , testCase "empty array" $ encode (Array mempty) @?= Right (BC.pack "[]\n")
    , testCase "keys sorted by code point, two-space indent" $
        encode (obj [("b", Number 2), ("a", Number 1), ("Z", Bool True)])
          @?= Right (BC.pack "{\n  \"Z\": true,\n  \"a\": 1,\n  \"b\": 2\n}\n")
    , testCase "nested containers, one element per line" $
        encode (obj [("xs", Array (V.fromList [Number 1, obj [("k", Null)]]))])
          @?= Right (BC.pack "{\n  \"xs\": [\n    1,\n    {\n      \"k\": null\n    }\n  ]\n}\n")
    , testCase "escaping: quote, backslash, named controls, other controls, raw non-ASCII" $
        encode (String (T.pack "a\"b\\c\nd\t\SOH\233\x2013"))
          @?= Right expectedEscaped
    , testCase "non-integer numbers are refused" $
        case encode (Number 1.5) of
          Left _ -> pure ()
          Right b -> fail ("accepted a float: " <> BC.unpack b)
    , testProperty "round trip: parse . encode = id, and encode . parse . encode = encode" $
        forAll genValue $ \v -> case encode v of
          Left e -> counterexample e False
          Right bytes -> case decodeStrict bytes of
            Nothing -> counterexample ("unparseable: " <> BC.unpack bytes) False
            Just v' -> v' === v .&&. encode v' === Right bytes
    ]
  where
    obj = Object . KeyMap.fromList . map (\(k, x) -> (Key.fromString k, x))
    -- U+00E9 and U+2013 stay raw UTF-8; the expected bytes are written out
    -- explicitly so the test does not compute them with the code under test.
    expectedEscaped = BC.pack "\"a\\\"b\\\\c\\nd\\t\\u0001" <> BC.pack "\xC3\xA9\xE2\x80\x93" <> BC.pack "\"\n"

-- | Values with integer numbers only, modest depth, and keys drawn from a
-- set that exercises code-point ordering (upper before lower, digits first).
genValue :: Gen Value
genValue = sized go
  where
    -- Fan-out is bounded per level, or the generator's size blows up
    -- exponentially with depth (it did: the first run never finished).
    go n
      | n <= 0 = leaf
      | otherwise =
          frequency
            [ (3, leaf)
            , (1, Array . V.fromList <$> few (go (n `div` 3)))
            , (2, obj <$> few ((,) <$> genKey <*> go (n `div` 3)))
            ]
    few g = choose (0, 4 :: Int) >>= \k -> vectorOf k g
    leaf =
      oneof
        [ pure Null
        , Bool <$> arbitrary
        , Number . fromInteger <$> arbitrary
        , String . T.pack <$> listOf genChar
        ]
    genKey = T.pack <$> listOf1 (elements "09AZaz_-.")
    genChar = frequency [(8, elements (['a' .. 'z'] ++ "\"\\ ")), (1, elements "\n\r\t\b\f\SOH\US"), (1, elements "\233\x2013\x1F600")]
    obj = Object . KeyMap.fromList . map (\(k, x) -> (Key.fromText k, x))

File: core/tests/canonical.rs
//! The canonical printer: exact bytes on fixed cases, and a round trip on
//! generated values (encode, parse, re-encode: same bytes; and the parse is
//! the original value). A port of the prototype's Test.Canonical, with the
//! same expected bytes written out rather than computed.

use rue_core::json::canonical::{encode, CanonicalError};
use serde_json::{json, Value};

#[test]
fn empty_object_and_array() {
    assert_eq!(encode(&json!({})).unwrap(), b"{}\n");
    assert_eq!(encode(&json!([])).unwrap(), b"[]\n");
}

#[test]
fn keys_sorted_by_code_point_two_space_indent() {
    let v = json!({"b": 2, "a": 1, "Z": true});
    assert_eq!(
        encode(&v).unwrap(),
        b"{\n  \"Z\": true,\n  \"a\": 1,\n  \"b\": 2\n}\n"
    );
}

#[test]
fn nested_containers_one_element_per_line() {
    let v = json!({"xs": [1, {"k": null}]});
    assert_eq!(
        encode(&v).unwrap(),
        b"{\n  \"xs\": [\n    1,\n    {\n      \"k\": null\n    }\n  ]\n}\n"
    );
}

#[test]
fn escaping_quote_backslash_named_controls_other_controls_raw_non_ascii() {
    let v = Value::String("a\"b\\c\nd\t\u{1}\u{e9}\u{2013}".to_string());
    // U+00E9 and U+2013 stay raw UTF-8; written out so the test does not
    // compute them with the code under test.
    let mut expected: Vec<u8> = b"\"a\\\"b\\\\c\\nd\\t\\u0001".to_vec();
    expected.extend_from_slice(&[0xC3, 0xA9, 0xE2, 0x80, 0x93]);
    expected.extend_from_slice(b"\"\n");
    assert_eq!(encode(&v).unwrap(), expected);
}

#[test]
fn non_integer_numbers_are_refused_with_their_path() {
    let v = json!({"a": [1, {"b": 1.5}]});
    assert_eq!(
        encode(&v),
        Err(CanonicalError::NonInteger {
            path: "$.a[1].b".to_string()
        })
    );
}

#[test]
fn round_trip_over_generated_values() {
    // xorshift32: deterministic across platforms, replayable from the seed.
    struct Rng(u32);
    impl Rng {
        fn next(&mut self) -> u32 {
            let mut x = self.0;
            x ^= x << 13;
            x ^= x >> 17;
            x ^= x << 5;
            self.0 = x;
            x
        }
        fn below(&mut self, n: u32) -> u32 {
            self.next() % n
        }
    }
    // Keys drawn from a set that exercises code-point ordering (digits, upper,
    // lower, punctuation); strings from letters, the two escaped ASCII
    // characters, named and unnamed controls, and raw non-ASCII.
    const KEY_CHARS: &[char] = &['0', '9', 'A', 'Z', 'a', 'z', '_', '-', '.'];
    const STR_CHARS: &[char] = &[
        'a',
        'q',
        'z',
        '"',
        '\\',
        ' ',
        '\n',
        '\r',
        '\t',
        '\u{8}',
        '\u{c}',
        '\u{1}',
        '\u{1f}',
        '\u{e9}',
        '\u{2013}',
        '\u{1f600}',
    ];
    fn gen(rng: &mut Rng, depth: u32) -> Value {
        let leaf = depth == 0 || rng.below(2) == 0;
        if leaf {
            match rng.below(4) {
                0 => Value::Null,
                1 => Value::Bool(rng.below(2) == 0),
                2 => Value::from(rng.next() as i64 - (u32::MAX / 2) as i64),
                _ => {
                    let n = rng.below(6);
                    Value::String(
                        (0..n)
                            .map(|_| STR_CHARS[rng.below(STR_CHARS.len() as u32) as usize])
                            .collect(),
                    )
                }
            }
        } else if rng.below(3) == 0 {
            let n = rng.below(5);
            Value::Array((0..n).map(|_| gen(rng, depth - 1)).collect())
        } else {
            let n = rng.below(5);
            let mut m = serde_json::Map::new();
            for _ in 0..n {
                let klen = 1 + rng.below(3);
                let k: String = (0..klen)
                    .map(|_| KEY_CHARS[rng.below(KEY_CHARS.len() as u32) as usize])
                    .collect();
                m.insert(k, gen(rng, depth - 1));
            }
            Value::Object(m)
        }
    }
    let mut rng = Rng(0x9E37_79B9);
    for _ in 0..500 {
        let v = gen(&mut rng, 3);
        let bytes = encode(&v).unwrap_or_else(|e| panic!("{e} for {v}"));
        let back: Value = serde_json::from_slice(&bytes)
            .unwrap_or_else(|e| panic!("unparseable ({e}): {}", String::from_utf8_lossy(&bytes)));
        assert_eq!(back, v);
        assert_eq!(encode(&back).unwrap(), bytes);
    }
}
